### PR TITLE
feat: background worker routines to shutdown client for migration

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description

This pr adds three background threads which can trigger the zeta-client to shutdown to help the process of migration
Thread 1 : Check for keygen and restart if new keygen is set 
Thread 2 : Check TSS list and restart if a new tss is added 
Thread 3 : Check current TSS , and restart if it is updated 

Closes https://github.com/zeta-chain/node/issues/2455
https://github.com/zeta-chain/node/issues/2492

# How Has This Been Tested?

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

## Summary by CodeRabbit

- **New Features**
	- Introduced structured background processing for handling TSS updates and key generation.
	- Added a new `Logger()` method to access logging capabilities within the application context.
	- Implemented a new method `GetTSS` to retrieve TSS objects directly from the observer client.
	- Added a constant backoff strategy for retry mechanisms.
  
- **Improvements**
	- Enhanced gRPC connection security.
	- Improved error handling and logging during shutdown processes.
	- Updated restart policies in Docker configuration to ensure service resilience.

- **Bug Fixes**
	- Addressed issues with task cancellation and shutdown sequence management.

### CODE REVIEW

File: cmd/zetaclientd/start.go
Comment (fbac) [inline cmd/zetaclientd/start.go:217]: This could be refined so we reduce the codebase:
```suggestion
	ctxwc, cancel := signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGTERM)
```

So we can remove all the code regarding the sig channel, saving resources and logic.

And replace this:
```go
	startLogger.Info().Msgf("awaiting shutdown signals")
	select {
	case <-notifyCtx.Done():
		cause := context.Cause(notifyCtx)
		startLogger.Info().Msgf("shutdown signal received , cause : %s", cause)

	case sig := <-ch:
		startLogger.Info().Msgf("stop signal received: %s", sig)
	}

	//stop chain observers
	for _, observer := range observerMap {
		observer.Stop()
	}
	zetacoreClient.Stop()

	return nil
```

for this leaner code:
```go
	startLogger.Info().Msgf("zetaclientd is running")
	<-notifyCtx.Done()

        startLogger.Info().Msgf("zetaclientd shutting down")
	for _, observer := range observerMap {
		observer.Stop()
	}

	zetacoreClient.Stop()

	return nil
```

File: zetaclient/zetacore/background_threads.go
Comment (fbac) [inline zetaclient/zetacore/background_threads.go:49]: what is the use of this line? Not sure if I'm missing something but we don't return `tss` so this line seems to not be doing anything.

File: zetaclient/zetacore/background_threads.go
Comment (fbac) [inline zetaclient/zetacore/background_threads.go:100]: Same here, this line seems disposable.

File: zetaclient/zetacore/background_threads.go
Comment (fbac) [inline zetaclient/zetacore/background_threads.go:152]: As well this seems disposable:
```suggestion
				logger.Info().Msgf("got new keygen at block %d", keygenUpdated.BlockNumber)
```

File: cmd/zetaclientd/start.go
Comment (lumtis) [inline cmd/zetaclientd/start.go:216]:
```suggestion
	// Create a notification context for background threads.
	// These threads are responsible for sending shutdown signals to the main thread.
```

File: cmd/zetaclientd/start.go
Comment (lumtis) [inline cmd/zetaclientd/start.go:220]: Why using defer here?

File: cmd/zetaclientd/start.go
Comment (lumtis) [inline cmd/zetaclientd/start.go:442]: I think "routine" is a better terminology than "thread" for Go

File: zetaclient/zetacore/background_threads.go
Comment (lumtis) [inline zetaclient/zetacore/background_threads.go:1]: `backgroud_threads` would be too abstract for a name
We should have filename reflecting content
- tss_update_listeners
- etc..

File: zetaclient/zetacore/background_threads.go
Comment (lumtis) [inline zetaclient/zetacore/background_threads.go:15]: Name is not reflective, it doesn't handle a TSS update, just wait for a TSS update
- ListenTSSUpdate
- WaitForTSSUpdate

File: zetaclient/zetacore/background_threads.go
Comment (lumtis) [inline zetaclient/zetacore/background_threads.go:63]: Same comment as above

**Review by lumtis — CHANGES_REQUESTED** [review]
Would be good to have a review from @swift1337 for this PR. 🙏

File: cmd/zetaclientd-supervisor/main.go
Comment (swift1337) [reply cmd/zetaclientd-supervisor/main.go]: Also, let's not log errors twice. 
```go
if err := cmd.Run(); err != nil {
				return errors.Wrap(err, "zetaclient process exited with error")
}
```

....


```go
if err := eg.Wait(); err != nil {
    logger.Error().Err(err).Msg("error while waiting")
}
```

File: zetaclient/orchestrator/orchestrator.go
Comment (swift1337) [inline zetaclient/orchestrator/orchestrator.go:120]: Could you explain this change?

File: zetaclient/zetacore/background_threads.go
Comment (kingpinXD) [reply zetaclient/zetacore/background_threads.go:15]: The reason for adding the `handle` keyword is the fact that when the Update happens, this routine triggers a shutdown for the main thread, using the `cancel()` function of the parent context.
Happy to change it if you still feel otherwise , resolving the other similar comments for now

File: zetaclient/context/app.go
Comment (swift1337) [inline zetaclient/context/app.go:73]: Let's not expose context's logger to the outer scope.

File: cmd/zetaclientd/start.go
Comment (kingpinXD) [reply cmd/zetaclientd/start.go]: The defer is to stop the background threads in case of a shutdown via os signals. I don't think its mandatory to have this , but can be good to have

File: zetaclient/zetacore/background_threads.go
Comment (swift1337) [inline zetaclient/zetacore/background_threads.go]: - Let's create a `zetaclient/maintenance` package and move TSS chores there. `zetaclient/zetacore/client.go` should be a client (i.e. req/res only), not a worker (there's 1 worker there atm which we need to move to orchestrator).
- Naming:
   - `(*TSSChore).WaitForUpdate(ctx) error` - blocking call. Returns `nil` on change
   - `(*TSSChore).WaitForNewTSSGeneration(ctx) error` - blocking call. Returns `nil` on change
   - `(*TSSChore).WaitForNewKeygen(ctx) error` - blocking call. Returns `nil` on change

**Review by swift1337 — CHANGES_REQUESTED** [review]

File: zetaclient/zetacore/background_threads.go
Comment (swift1337) [reply zetaclient/zetacore/background_threads.go]: Also, the logic with cancel function is confusing, it's hard to grasp who actually calls cancel. I would not use `bg.WithCancel()` right now as well as this: (start.go)
```go
	// Create a notification context for background threads.These threads are responsible for sending shutdown signals to the main thread.
	notifyCtx, cancel := context.WithCancelCause(ctx)

	// Start background threads
	defer startBackgroundThreads(notifyCtx, cancel, zetacoreClient, masterLogger)
```

How I envision it:
```go
tssChores := maintenance.NewTSSChores(func(){
   // callback that is being invokes if we detect any changes in `Wait*` calls
   sigChannel <-  syscall.SIGINT
})


// invokes bg task
tssChores.Start(ctx)

```

and then we can simply implement bg opt `bg.OnComplete(callback)`








**Note**: Orchestrator is responsible for (de)provisioning observers/signers, no need to manually stop observers in start.go as it should happen within orchestrator's lifecycle.

File: cmd/zetaclientd/start.go
Comment (swift1337) [reply cmd/zetaclientd/start.go:329]: oh, I see that it haven't been merged with the latest `develop`. I fixed this

File: zetaclient/orchestrator/orchestrator.go
Comment (kingpinXD) [reply zetaclient/orchestrator/orchestrator.go:120]: the observervers are already being stopped here 
https://github.com/zeta-chain/zeta-node/blob/59515eb811ab940564b336cf08a5ac328986ebeb/cmd/zetaclientd/start.go#L356-L358

Comment (swift1337) [reply zetaclient/orchestrator/orchestrator.go:120]: https://github.com/zeta-chain/node/pull/2538#discussion_r1689904775

File: zetaclient/zetacore/background_threads.go
Comment (kingpinXD) [reply zetaclient/zetacore/background_threads.go]: I agree that using bg.Cancel() is not the cleanest way to do this, but the arch mentioned in this comment should be part of a separate refactor .
In my mind , background tasks is fine , as a package , We should ideally have two types of tasks , ones that read only and ones that write something back to the parent thread (naming can be decided later ), that way it can be more general , rather that adding something like `TSSChores` which is too specific .

This PR tries to keep changes to the zetaclient minimal .

File: cmd/zetaclientd/start.go
Comment (kingpinXD) [reply cmd/zetaclientd/start.go]: it should stop the thread here 
https://github.com/zeta-chain/zeta-node/blob/2cce32c2fefc3ef7be37c797a47d3a6cd9a51905/pkg/bg/bg.go#L57-L59

I know its not the cleanest solution , and planning to implement a bigger refactor for background threads .
https://github.com/zeta-chain/node/pull/2538#discussion_r1689939726

Wil add a task once I finalize the details

File: zetaclient/context/app.go
Comment (kingpinXD) [reply zetaclient/context/app.go:73]: is there an issue you see with that , 
I am using it here https://github.com/zeta-chain/zeta-node/blob/2cce32c2fefc3ef7be37c797a47d3a6cd9a51905/zetaclient/zetacore/client_tss_migration_listners.go#L55-L55

File: cmd/zetaclientd/start.go
Comment (kingpinXD) [reply cmd/zetaclientd/start.go:215]: I think , we would have to define a custom error , to fetch the cause in this case .
I want to use the cause to print out which job sends the shutdown request . 
What do you think ?

Comment (fbac) [reply cmd/zetaclientd/start.go:215]: In the case of an error that would be printed anyway, right? as we execute the `bg.Work` routines with `WithLogger` and `WithName` the first error log we have before "zetaclientd shutting down" we can assume it's the one calling `cancel()`.

Another way to do it, more deterministic, is to modify the context that we create in StartTssMigrationRoutines and return it along with the cancelCauseFunc:
```go
	causeCtx, causeCancel := context.WithCancelCause(ctx)
	bg.Work(
		migrationRoutinesContext,
		c.HandleTSSUpdate,
		bg.WithName("HandleTSSUpdate"),
		bg.WithLogger(masterLogger),
		bg.WithCallback(causeCancel),
...

return migrationRoutinesContext, cancel
```
When the first task fails will cancel causeCtx:
- all the other tasks should receive a ctx.Done() and close.
- the `causeCtx` would carry the error and we can recover it from the main goroutine.

This would require some prior testing, as I'm just talking from memory and haven't tested it myself.

File: zetaclient/zetacore/background_threads.go
Comment (kingpinXD) [reply zetaclient/zetacore/background_threads.go]: Summarizing points discussed and additional thoughts . From what I understand, there are two main things that you are suggesting 
- Use a callback instead of the cancel function: I agree that cancel function is not the ideal option, but it provides some advantages, such as canceling all sub-routines when the parent is canceled. The cancelFunction provided here is actually the callback (renamed so that the terms make more sense ) 

- Made the functions as part of a separate struct and not `client` :  I don't see a lot of advantage of doing this , considering we are the functions that these routines use are all part of the client right now . Unless we are planning to modify functions such as `GetTss(ctx)` to `GetTss(ctx,client)`

File: cmd/zetaclientd/start.go
Comment (kingpinXD) [reply cmd/zetaclientd/start.go:215]: I think your first point makes sense , we are printing inside the `Work` function , I can remove the cancelCause

**Comment by kingpinXD** [comment]
Moving to draft for now , TSS migration test is broken on current develop . Will be fixing that first

**Comment by kingpinXD** [comment]
> Moving to draft for now , TSS migration test is broken on current develop . Will be fixing that first

The pr is ready for a review

File: changelog.md
Comment (lumtis) [inline changelog.md:39]: To be moved to unreleased

File: zetaclient/zetacore/client_query_observer.go
Comment (lumtis) [inline zetaclient/zetacore/client_query_observer.go:170]: Can we please add comment

Comment (lumtis) [inline zetaclient/zetacore/client_query_observer.go:170]: I think we have a test suite for `client_query` and we can add unit test for this function

**Comment by lumtis** [comment]
Let's wait on https://github.com/zeta-chain/node/pull/2560 before considering merging this one

File: cmd/zetaclientd/start.go
Comment (swift1337) [inline cmd/zetaclientd/start.go:362]: Latest `develop` doesn't have observer.Stop(), only `zetacoreClient.Stop()` https://github.com/zeta-chain/node/blob/develop/cmd/zetaclientd/start.go#L350. Please make sure this is up-to date with develop

File: zetaclient/zetacore/background_threads.go
Comment (swift1337) [reply zetaclient/zetacore/background_threads.go]: Several points from my side:
- By implementing reusable code components, we can simplify the mental model behind the project.
- In order to keep the codebase maintainable, we need to reduce the technical debt or at least stop increasing it. In my point of view, adding net-new worker logic in the client's package (a package that should handle req/res only) is indeed a technical debt increase due to making a "shortcut."
- `bg` package can and should be improved in the future PRs, but for this PR,`bg.OnSuccess(func())` is more than enough (IMO)

Comment (swift1337) [reply zetaclient/zetacore/background_threads.go]: I implemented proposed changes, feel free to ask any questions:
- https://github.com/zeta-chain/node/pull/2538/commits/f993da5aa5db578556741711bbab7f9ae8b732b7
- https://github.com/zeta-chain/node/pull/2538/commits/7c113a5059da6ce9e6e7a87dc4c1fcecaeacbfe8

TLDR:
```go
// cmd/zetaclient/start.go
func start(_ *cobra.Command, _ []string) error {
  // ...
  
  // Creating a channel to listen for os signals (or other signals)
  signalChannel := make(chan os.Signal, 1)
  signal.Notify(signalChannel, syscall.SIGINT, syscall.SIGTERM)
  
  // Maintenance workers ============
  maintenance.NewTSSListener(zetacoreClient, masterLogger).Listen(ctx, func() {
      masterLogger.Info().Msg("TSS listener received an action to shutdown zetaclientd.")
      signalChannel <- syscall.SIGTERM
  })
  
  sig := <-signalChannel
  startLogger.Info().Msgf("Stop signal received: %q", sig)
  
  zetacoreClient.Stop()
}
```

cc @lumtis

**Review by swift1337 — APPROVED** [review]
LGTM

**Review by lumtis — APPROVED** [review]
Code organization and implementation looks good to me, I guess this can be tested E2E in the TSS migration test?

**Comment by kingpinXD** [comment]
> Code organization and implementation looks good to me, I guess this can be tested E2E in the TSS migration test?

Yes , the TSS migration test would use these routines

**Pull request merged**
